### DISCUSSION TITLE
iOS16.0 visible wrong in pageSource

### DISCUSSION THREAD

**Discussion by RSM-SBA** [appium/appium#17599] [Q&A]
I only want to ask if somebody has also noticed that within an iOS16.0 simulator, sometimes elements are visible=false in the pageSource, but they are shown and not overlayed by anything in the app.

**Comment by jlipps** [comment]
Dunno, but visibility checks are best-effort and known to sometimes be unreliable due to misreporting from the Apple side

**Comment by anupamg-discovery** [comment]
@RSM-SBA Have you got any solution for this ? I am also struggling with this.

**Comment by nanoscopic** [comment]
Appium uses WDA. WDA uses private Apple APIs to get the element structure.

That element structure? Misreports visibility. It is a bug in the underlying Apple system.

Even worse there is no direct way through xctest ( which wda uses ) to get the current vertical scroll position within an app, so you can't even reasonably tell if an item is visible via it's coordinates in all cases. ( I've seen bugs around that... )

I suspect the solution to this is to utilize the "debug view hierarchy", which is a RSD accessible call to get element structure that does not use WDA or xctest at all. It does require the ability to attach a debugger to your app though. And Appium ( nor any open source ) has yet implemented "debug view hierarchy". It returns the exact set of visible elements and even includes images of each element.
